backend-docker/pier/db/db.go

```diff
@@ -0,0 +1,551 @@
+	ExitCode      int
+	ConsoleOutput string
+}
+
+// TaskTable contains tasks related to a specific job (used to store data in a database)
+type TaskTable struct {
+	ID            string
+	Name          string
+	ContainerID   string
+	Error         string
+	ExitCode      int
+	ConsoleOutput string
+	Revision      int
+	JobID         string
+}
+
+// LatestOutput used to serialize console output to JSON
+type LatestOutput struct {
+	Name          string
+	ConsoleOutput string
+}
+
+// Bind describes the binding between an IOPort and a docker volume
+type Bind struct {
+	IOPort   IOPort
+	VolumeID dckr.VolumeID
+}
+
+// Service describes metadata for a GEF service (used to serialize JSON)
+type Service struct {
+	ID          ServiceID
+	ImageID     dckr.ImageID
+	Name        string
+	RepoTag     string
+	Description string
+	Version     string
+	Created     time.Time
+	Size        int64
+	Input       []IOPort
+	Output      []IOPort
+}
+
+// ServiceTable describes metadata for a GEF service (used to store data in a database)
+type ServiceTable struct {
+	ID          string
+	ImageID     string
+	Name        string
+	RepoTag     string
+	Description string
+	Version     string
+	Revision    int
+	Created     time.Time
+	Size        int64
+}
+
+// ServiceID exported
+type ServiceID string
+
+// IOPort is an i/o specification for a service
+// The service can only read data from volumes and write to a single volume
+// Path specifies where the volumes are mounted
+type IOPort struct {
+	ID   string
+	Name string
+	Path string
+}
+
+// IOPortTable is used to store data in a database
+type IOPortTable struct {
+	ID        string
+	Name      string
+	Path      string
+	IsInput   bool
+	Revision  int
+	ServiceID string
+}
+
+// InitDb initializes the database engine
+func InitDb() (Db, error) {
+	var dataBaseHandler Db
+	dataBase, err := sql.Open("sqlite3", SQLiteDataBasePath)
+	if err != nil {
+		return dataBaseHandler, err
+	}
+	// For each table GORP has a special field to keep information about data version
+	dataBaseMap := &gorp.DbMap{Db: dataBase, Dialect: gorp.SqliteDialect{}}
+	dataBaseMap.AddTableWithName(JobTable{}, "Jobs").SetKeys(false, "ID").SetVersionCol(GorpVersionColumn)
+	dataBaseMap.AddTableWithName(TaskTable{}, "Tasks").SetKeys(false, "ID").SetVersionCol(GorpVersionColumn)
+	dataBaseMap.AddTableWithName(ServiceTable{}, "Services").SetKeys(false, "ID").SetVersionCol(GorpVersionColumn)
+	dataBaseMap.AddTableWithName(IOPortTable{}, "IOPorts").SetVersionCol(GorpVersionColumn)
+	err = dataBaseMap.CreateTablesIfNotExists()
+	dataBaseHandler = Db{*dataBaseMap}
+	return dataBaseHandler, err
+}
+
+// AddJob adds a job to the database
+func (d *Db) AddJob(job Job) error {
+	storedJob := d.MapJSON2StoredJob(job)
+	return d.Insert(&storedJob)
+}
+
+// RemoveJob removes a job and all corresponding tasks from the database
+func (d *Db) RemoveJob(jobID JobID) error {
+	_, err := d.Exec("DELETE FROM Tasks WHERE jobID=?", string(jobID))
+	if err == nil {
```

Comment:
it's more common to test the error and return if it's not null, so the error cases get out of the way:
`if err != nil { return err }`